### PR TITLE
Pythonic implementation of LRU Cache

### PR DESCRIPTION
**Pull request by yatharthmathur**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: other/lru_cache_pythonic.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+The following implementation of LRU Cache is one of the most elegant pythonic implementations. 
+It only uses the in-built python dictionary (https://docs.python.org/3/library/stdtypes.html#typesmapping). This works because 
+the Python dictionary maintains the order of insertion of keys and ensures O(1) operations on insert, delete and access.
+'''
+class LRUCache(dict):
+    '''
+    Initialize an LRU Cache with given capacity.
+    capacity : int -> the capacity of the LRU Cache
+    '''
+    def __init__(self, capacity : int):
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py]: fixed in commit fc53f3b6247ec1fda1c64b71001767645652ff78

File: other/lru_cache_pythonic.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    returns -> value : Any -> any object that is stored as a value inside of the LRU cache.
+
+    >>> cache = LRUCache(2)
+    >>> cache.put(1,1)
+    >>> cache.get(1)
+    1
+    >>> cache.get(2)
+    KeyError: '2 not found.'
+    '''
+    def get(self, key:Hashable)->Any:
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py:28]: fixed in commit fc53f3b6247ec1fda1c64b71001767645652ff78

File: other/lru_cache_pythonic.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        val = self.pop(key) # Pop the key-value and re-insert to maintain the order
+        self[key] = val
+        return val
+
+    '''
+    This method puts the value associated with the key provided inside of the LRU cache.
+    key : Hashable -> a hashable object that is mapped to a value inside of the LRU cache.
+    value: Any -> any object that is to be associated with the key inside of the LRU cache.
+    '''
+    def put(self, key:Hashable, value:Any):
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py]: fixed in commit fc53f3b6247ec1fda1c64b71001767645652ff78

File: other/lru_cache_pythonic.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            self.pop(key)
+            self[key] = value
+            return
+
+        if self.remaining > 0: self.remaining -= 1
+        # To pop the least recently used item from the dictionary 
+        else: self.pop(next(iter(self)))
+        self[key] = value
+
+def main():
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py]: fixed in commit fc53f3b6247ec1fda1c64b71001767645652ff78

File: other/lru_cache_pythonic.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    returns -> value : Any -> any object that is stored as a value inside of the LRU cache.
+
+    >>> cache = LRUCache(2)
+    >>> cache.put(1,1)
+    >>> cache.get(1)
+    1
+    >>> cache.get(2)
+    KeyError: '2 not found.'
+    '''
+    def get(self, key:Hashable)->Any:
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py:19]: fixed in b237fb3d0bfdf3b609cda185032a95814d42f447

File: other/lru_cache_pythonic.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    value: Any -> any object that is to be associated with the key inside of the LRU cache.
+    >>> cache = LRUCache(2)
+    >>> cache.put(3,3)
+    >>> cache
+    {3:3}
+    >>> cache.put(2,2)
+    >>> cache
+    {3:3, 2:2}
+    '''
+    def put(self, key:Hashable, value:Any)->None:
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py:41]: fixed in b237fb3d0bfdf3b609cda185032a95814d42f447

File: other/lru_cache_pythonic.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            self.pop(key)
+            self[key] = value
+            return
+
+        if self.remaining > 0: self.remaining -= 1
+        # To pop the least recently used item from the dictionary 
+        else: self.pop(next(iter(self)))
+        self[key] = value
+
+def main()->None:
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py:65]: fixed in b237fb3d0bfdf3b609cda185032a95814d42f447

File: other/lru_cache_pythonic.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
```
[the remark this answers is not in the document]
Comment (yatharthmathur) [reply other/lru_cache_pythonic.py:65]: Main method : doesn't require any doc tests.

**Review by poyea — COMMENTED** [review]
Why this is more Pythonic? Is there any reason not to modify the current implementations: [LRU](https://github.com/TheAlgorithms/Python/blob/master/other/lru_cache.py), [least_recently_used](https://github.com/TheAlgorithms/Python/blob/master/other/least_recently_used.py)

**Comment by cclauss** [comment]
@yatharthmathur Do you know how to rebase this PR?

**Review by CaedenPH — COMMENTED** [review]
I like the idea of this pure python implementation of lru_cache, and the code is doctested and looks good to me. 
However, this pr has had no response for 2 years so I think it should be closed @cclauss @tianyizheng02
@yatharthmathur feel free to re-implement this algorithm in a new pull request

**Review by cclauss — APPROVED** [review]
Let's do this

**Comment by yatharth-pdgt** [comment]
Hey @cclauss 
Sorry, I missed your comment. I see we have merged the pull request! That's great! Thanks for your consideration.
cc: @CaedenPH

**Comment by yatharthmathur** [comment]
Hey @cclauss
Sorry, I missed your comment. I see we have merged the pull request! That's great! Thanks for your consideration.
cc: @CaedenPH

**Pull request merged**